Parameter improvements to Cluster Health API wait for shards

abeyad opened the pull request:
Previously, the cluster health API used a strictly numeric value for `wait_for_active_shards`. However, with the introduction of ActiveShardCount and the removal of write consistency level for replication operations, `wait_for_active_shards` is used for write operations to represent values for ActiveShardCount. This PR moves the cluster health API's usage of `wait_for_active_shards` to be consistent with its usage in the write operation APIs.

This PR also changes `wait_for_relocating_shards` from a numeric value to a simple boolean value `wait_for_no_relocating_shards` to set whether the cluster health operation should wait for
all relocating shards to complete relocation.

Breaking changes summary:
 1. `wait_for_active_shards`: this is *not* a breaking change because the same numeric values that were allowed before are allowed now.  We have just added the ability to specify `all` as well, in conformity with the the write operation APIs that also use `wait_for_active_shards`
 2. `wait_for_no_relocating_shards`: this *is* a breaking change because it differs from the old name `wait_for_relocating_shards` and it takes a boolean value to represent whether we should wait for the cluster to have no relocations, whereas the old `wait_for_relocating_shards` took a number to represent how many relocating shards to wait on (if this value was set, it was pretty much always set to `0`)

Closes #20216

abeyad commented:
@elasticmachine retest this please

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthRequest.java:
```diff
@@ -102,24 +103,52 @@ public ClusterHealthRequest waitForYellowStatus() {
-        this.waitForRelocatingShards = waitForRelocatingShards;
+    /**
+     * Sets whether the request should wait for there to be no relocating shards before
+     * retrieving the cluster health status.  Defaults to <code>false</code>, meaning the
```
`{@code false}`?

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthRequest.java:
```diff
@@ -102,24 +103,52 @@ public ClusterHealthRequest waitForYellowStatus() {
-    public ClusterHealthRequest waitForActiveShards(int waitForActiveShards) {
-        this.waitForActiveShards = waitForActiveShards;
+    /**
+     * Sets the number of shard copies that must be active before getting the health status.
```
"Sets the number of shard copies that must be active *across all indices* before getting the health status"?

abeyad replied:
done

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthRequest.java:
```diff
@@ -102,24 +103,52 @@ public ClusterHealthRequest waitForYellowStatus() {
+     * Set this value to {@link ActiveShardCount#ALL} to wait for all shards (primary and
+     * all replicas) to be active across all indices in the cluster. Otherwise, use
+     * {@link ActiveShardCount#from(int)} to set this value to any non-negative integer, up to the
+     * total number of shard copies that would exist across all indices in the cluster.
```
s/that would exist across all indices in the cluster./to wait for./?

abeyad replied:
done

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthRequest.java:
```diff
@@ -102,24 +103,52 @@ public ClusterHealthRequest waitForYellowStatus() {
+     * shard count is passed in, instead of having to first call {@link ActiveShardCount#from(int)}
+     * to get the ActiveShardCount.
+     */
+    public ClusterHealthRequest waitForActiveShards(final int waitForActiveShards) {
```
You even kept the java api compatible here. Very kind of you.

abeyad replied:
😀

nik9000 commented on docs/reference/cluster/health.asciidoc:
```diff
@@ -82,14 +82,14 @@ The cluster health API accepts the following request parameters:
-    A number controlling to how many active
-    shards to wait for. Defaults to not wait.
+    A number controlling to how many active shards to wait for, or `all` to wait
+    for all shards in the cluster to be active. Defaults to not wait.
```
"or `none` to not wait. Defaults to `none`."?

abeyad replied:
done

nik9000 commented on rest-api-spec/src/main/resources/rest-api-spec/test/cluster.health/20_request_timeout.yaml:
```diff
@@ -16,3 +16,22 @@
+  - do:
+      catch: request_timeout
+      cluster.health:
+        timeout: 1s
```
`1ms`?

nik9000 commented:
Left a few minor things, mostly documentation wording suggestions. LGTM.

abeyad commented:
@nik9000 thanks for the review! i pushed https://github.com/elastic/elasticsearch/pull/20223/commits/b50e472dd053b078f3fee8c73535063944fe00fc that addresses your comments

nik9000 commented:
LGTM

clintongormley commented:
I'm OK with this change, but I would add a check for `wait_for_relocating_shards` and throw an exception if present, otherwise this parameter will just be silently ignored.

it'd be better to make REST param parsing strict (https://github.com/elastic/elasticsearch/issues/14719) rather than dealing with these on a case-by-case basis

abeyad commented:
@clintongormley I pushed a commit to check for the old param and throw an exception: https://github.com/elastic/elasticsearch/pull/20223/commits/e4e946b4566f36725f76c3baca3674ba99aa134d

I agree that it would be good to have a general solution to making REST param parsing strict

niemyjski commented:
So in our use case when we create an index we want to block client calls until all shards are created but we don't care about replicas. So I guess with this change I would need to specify `all`.  Personally, I'd like if it was more like cluster heath api and I could just wait for `yellow` as shard size might not be known... But `all` works for me.

abeyad commented:
@niemyjski you wouldn't need to specify anything, because in 5.0, by default index creation will not return until all of the primary shards are active (see https://github.com/elastic/elasticsearch/pull/18985).  

Suppose you had 2 replicas in addition to your primary.  If you wanted to wait for at least 1 replica to become active in addition to the primaries before proceeding with the write, then you would use `wait_for_active_shards=2` (1 for primary plus 1 for replica).  If you wanted to wait for all replicas, you would specify `wait_for_active_shards=3` or more easier `wait_for_active_shards=all`

niemyjski commented:
Awesome! I didn't get that from reading the docs, that's good to know!

bleskes commented:
> Awesome! I didn't get that from reading the docs, that's good to know!

@niemyjski if you have any suggestion on how to improve the docs, please let us know (just click that edit button on the website :))
